### PR TITLE
Clair parser: fixes TypeError: string indices must be integers

### PR DESCRIPTION
**Pull request by alles-klar**
Current parser is broken and throws Server Error 500 / "TypeError: string indices must be integers" when uploading a scan.

Also adding some unittests.

----
On behalf of DB Systel GmbH

### CODE REVIEW

**Review by damiencarol — CHANGES_REQUESTED** [review]

**Comment by damiencarol** [comment]
@alles-klar what bug are you trying to fix?
is your PR a draft (not finished)? please mark it as draft if so.

**Comment by alles-klar** [comment]
@damiencarol PR is ready from my perspective. Since https://github.com/DefectDojo/django-DefectDojo/pull/3827 the clair parser doesn't work at all and you get "TypeError: string indices must be integers".

File: dojo/unittests/tools/test_clair_parser.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from django.test import TestCase
+from dojo.tools.clair.parser import ClairParser
+
+
+class TestFile(object):
```
Comment (damiencarol) [inline dojo/unittests/tools/test_clair_parser.py:5]: this class is useless as you have files in the pull request

File: dojo/unittests/tools/test_clair_parser.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        findings = parser.get_findings(my_file_handle, None)
+        my_file_handle.close()
+        self.assertEqual(0, len(findings))
+
+    def test_many_findings(self):
+        my_file_handle = open("dojo/unittests/scans/clair/many_vul.json")
+        parser = ClairParser()
+        findings = parser.get_findings(my_file_handle, None)
+        my_file_handle.close()
+        self.assertEqual(35, len(findings))
```
[the remark this answers is not in the document]
Comment (alles-klar) [reply dojo/unittests/tools/test_clair_parser.py:19]: done

File: dojo/tools/clair/parser.py [hunk 3]
```diff
@@ -15,9 +15,22 @@ def get_description_for_scan_types(self, scan_type):
...
+    def parse_json(self, json_output):
+        try:
+            data = json_output.read()
+            try:
+                tree = json.loads(str(data, 'utf-8'))
+            except:
+                tree = json.loads(data)
+            subtree = tree.get('vulnerabilities')
+        except:
+            raise Exception("Invalid format")
```
[the remark this answers is not in the document]
Comment (alles-klar) [reply dojo/tools/clair/parser.py]: I removed the surrounding try/except. Is that what you mean?

[the remark this answers is not in the document]
Comment (valentijnscholten) [reply dojo/tools/clair/parser.py]: no, you are raising "Invalid format" which hides the original exception.

should be something like
try:
....
except as e:
raise Exception("Invalid format", e)

[the remark this answers is not in the document]
Comment (damiencarol) [reply dojo/tools/clair/parser.py]: or we can remove the try/catch better IMO

**Review by damiencarol — APPROVED** [review]

**Comment by damiencarol** [comment]
@alles-klar good job!

**Review by valentijnscholten — APPROVED** [review]

**Pull request merged**